### PR TITLE
Remove nmcli modify dependency on type parameter

### PR DESCRIPTION
**Pull request by trewest**
##### SUMMARY

This change introduces the ability to modify nmcli connections without setting type parameter

Fixes https://github.com/ansible-collections/community.general/issues/2858

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

nmcli

##### ADDITIONAL INFORMATION

Updates the nmcli module in order to support nmcli connection updates without setting the type parameter.
This change was in order to support updating IP routes for existing connections as follows:

Tasks in playbook:
```
 - name: Ensure connection {{ connection_name }} does not exist
      community.general.nmcli:
        conn_name: "{{ connection_name }}"
        state: absent

    - name: Setup IPv4 connection
      block:
        - name: Create connection {{ connection_name }}
          vars:
            address: "192.168.10.10/24"
            static_route: "10.0.10.0/24 192.168.10.1"
          community.general.nmcli:
            conn_name: "{{ connection_name }}"
            ifname: "{{ connection_name }}"
            type: ethernet
            ip4:
              - "{{ address }}"
            routes4:
              - "{{ static_route }}"
            method4: manual
            state: present

        - name: Update routes for {{ connection_name }}
          vars:
            static_route: "10.0.20.0/24 192.168.10.1"
          community.general.nmcli:
            conn_name: "{{ connection_name }}"
            routes4:
              - "{{ static_route }}"
            state: present

        - name: Update route metric
          vars:
            static_route: "10.0.20.0/24 192.168.10.1"
          community.general.nmcli:
            conn_name: "{{ connection_name }}"
            routes4:
              - "{{ static_route }}"
            route_metric4: 10
            state: present
```

Create a task to create an nmcli connection and then a subsequent task to modify the connection without a type parameter

Current behavior from main branch:
```
PLAY [playbook] ******************************************************************************************************************************************************************************************************************************

TASK [Ensure connection eno2 does not exist] *************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Create connection eno2] ****************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Update routes for eno2] ****************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Update route metric] *******************************************************************************************************************************************************************************************************************
ok: [localhost]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

Behavior after changes:
```
PLAY [playbook] ******************************************************************************************************************************************************************************************************************************

TASK [Ensure connection eno2 does not exist] *************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Create connection eno2] ****************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Update routes for eno2] ****************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Update route metric] *******************************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=4    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0  
```

### CODE REVIEW

File: changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - nmcli - Remove nmcli modify dependency on type parameter (https://github.com/ansible-collections/community.general/issues/2858).
```
Comment (felixfontein) [inline changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml:2]:
```suggestion
  - nmcli - remove nmcli modify dependency on ``type`` parameter (https://github.com/ansible-collections/community.general/issues/2858).
```

File: changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
```
Comment (felixfontein) [inline changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml:1]: To me this rather looks like a feature:
```suggestion
minor_changes:
```

File: changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml (hunk 1 above)
Comment (trewest) [reply changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml]: done

File: changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml (hunk 2 above)
Comment (trewest) [reply changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml]: done

**Comment by felixfontein** [comment]
CC @Ajpantuso @jwennerberg @jfpanisset @haddystuff as you recently worked on this module.

**Comment by felixfontein** [comment]
@trewest I think it would be good to also mention this in the documentation for the `type` option.

@Ajpantuso @jwennerberg @jfpanisset @haddystuff would be really great if you could take a look. If nobody comments on this by say end of next week I'll merge this so it can be inclued in the next set of releases.

**Comment by trewest** [comment]
So I was investigating this a little bit more and noticed that sometimes there are other required fields in order to update a connection as expected. Take for example a preexisting connection with IPv4 routes in which you wanted to run a task to update the `ipv4.route-metric`.

```
- name: Ensure connection {{ connection_name }} does not exist
  community.general.nmcli:
    conn_name: "{{ connection_name }}"
    state: absent

- name: Setup IPv4 connection
  when: addrfamily == 'ipv4'
  block:
      - name: Create connection {{ connection_name }}
        vars:
          address: "192.168.10.10/24"
          static_route: "10.0.10.0/24 192.168.10.1"
        community.general.nmcli:
          conn_name: "{{ connection_name }}"
          ifname: "{{ connection_name }}"
          type: ethernet
          ip4:
            - "{{ address }}"
          routes4:
            - "{{ static_route }}"
          method4: manual
          state: present
    
      - name: Update route metric
        community.general.nmcli:
          conn_name: "{{ connection_name }}"
          route_metric4: 10
          state: present
```
Instead of only modifying the `ipv4.route-metric` the last task will also empty the `ipv4.routes` list since the `routes4` parameter was not supplied:
```
➜  ~ nmcli -f ipv4.addresses,ipv4.routes,ipv4.route-metric connection show eno2
ipv4.addresses:                         192.168.10.10/24
ipv4.routes:                            --
ipv4.route-metric:                      10
```

I guess the question is, should type, ip4, and ip6 settings just be special fields that we will retrieve from the current connection if  they aren't supplied by the user or should the list be more extensive?

**Comment by jfpanisset** [comment]
I am not familiar with the rest of the code, but I was wondering why the need to remap `802-3-ethernet` to `ethernet` as the connection type, as the [nmcli man page](https://developer-old.gnome.org/NetworkManager/stable/nmcli.html) seems to indicate that `ethernet` is an alias for `802-3-ethernet` and either can be used? If the rest of the code depends on having `type` set to `ethernet`, should the mapping table also remap `802-11-wireless` to `wifi` and `802-11-olpc-mesh` to `olpc-mesh`?

**Comment by trewest** [comment]
Thanks @jfpanisset, yeah you were right. It looks like the aliases just hadn't been included in the `ip_conn_type` check. Added both `802-3-ethernet` and `802-11-wireless` but doesn't seem that `olpc-mesh` has been added so I have omitted `802-11-olpc-mesh` for now.

**Comment by trewest** [comment]
I would still like to learn from some of the other contributors if there is a use case for getting other fields from the current connection if they aren't supplied by the user (like mentioned in https://github.com/ansible-collections/community.general/pull/4108#issuecomment-1037731036). I'm not sure that it makes sense to only make an exception for `type` if all other parameters need to be supplied in order to update a connection.

File: plugins/modules/net_tools/nmcli.py [hunk 3]
```diff
@@ -1536,6 +1537,7 @@ def ip_conn_type(self):
             'team',
             'vlan',
             'wifi',
+            '802-11-wireless',
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:1540]: Do these changes need a mention in the changelog?

Comment (trewest) [reply plugins/modules/net_tools/nmcli.py:1540]: done

File: changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml [hunk 4]
```diff
@@ -1,2 +1,3 @@
 minor_changes:
   - nmcli - remove nmcli modify dependency on ``type`` parameter (https://github.com/ansible-collections/community.general/issues/2858).
+  - nmcli - add missing connection aliases '802-3-ethernet' and '802-11-wireless'
```
Comment (felixfontein) [inline changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml:3]:
```suggestion
  - nmcli - add missing connection aliases ``802-3-ethernet`` and ``802-11-wireless`` (https://github.com/ansible-collections/community.general/pull/4108).
```

Comment (trewest) [reply changelogs/fragments/4108-nmcli-support-modifcation-without-type-param.yml]: done

**Comment by felixfontein** [comment]
Any more comments on this? If nobody comments, I'll merge this by the end of next week.

**Comment by felixfontein** [comment]
@trewest thanks for your contribution!
@jfpanisset thanks for reviewing!

**Pull request merged**